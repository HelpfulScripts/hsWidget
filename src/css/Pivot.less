
.right { text-align:right; }
.scrolly { overflow-y: scroll; }


.pivot {
    width: 100%;
    border-bottom: 5px solid white;
    // height: 100%;
    & .mon_header {
        height: 1.4em;
        background-color: #66a;
        color: white;
    }
    & .row {
        width: 100%;
        padding:0;
        height: 1.4em;
    }
    & span {
        &.name { 
            width: 10%; 
            white-space: nowrap;
            padding-top: 3px;
            padding-left: 4px;
        }
        display: inline-block;
        width: 20%;
        font-size: 80%;
    }

    .mon0   { background-color: #99b; color: white; }
    .mon1   { background-color: #bbd; color: black; }
    .mon2   { background-color: #ddf; color: black; }
    .mon3   { background-color: #eee; color: black; }
    .mon4   { background-color: #fff; color: black; }
}