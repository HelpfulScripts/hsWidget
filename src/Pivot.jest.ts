import m from "mithril";
import { Pivot, Aggregator, PivotAttrs, ColumnGenerator }    from './Pivot';
const root = window.document.createElement("div");

const headers = ['color', 'shape', 'Value', 'area'];
const table = [
     ['blue',  'square',   17, 'NW'],
     ['red',   'circle',   37, 'N'],
     ['blue',  'triangle', 22, 'S'],
     ['green', 'square',   18, 'NE'],
     ['green', 'cone',      9, 'SE'],
     ['blue',  'circle',   12, 'S']
];

// ColumnAccess function: calculates the sum per row
const sums:ColumnGenerator = (namedColumns:{[name:string]:Aggregator}, value: number):void => {
     const colName = 'Sums';
     namedColumns[colName] = (<number>namedColumns[colName]||0) + value; 
}

describe('Pivot', () => {
    describe('sum', () => {
        beforeAll(()=>{
            m.mount(root, { view: () => m('.hs_white', [
                m(Pivot, <PivotAttrs>{ 
                      pivotName: 'click to expand rows',
                      table: {data: table, header:headers},
                      columns:[{Value:'>sums'}, {Value:'shape'}],
                      by: ['color', 'area']
                }),
             ])});
        });
    
        it('should match snapshot', () => {
            expect(root).toMatchSnapshot();
        });    
    });
    describe('max and generator', () => {
        beforeAll(()=>{
            m.mount(root, { view: () => m('.hs_white', [
                m(Pivot, <PivotAttrs>{ 
                      pivotName: 'by Area',
                      pivotHeaders: ['Sums', 'red', 'green', 'blue'],
                      table: {data: table, header:headers},
                      columns:[{'Value':sums}, {'Value':'>color'}], // use `ColumnAccess` function and string column
                      by: ['area', 'shape']
                }),
             ])});
        });
    
        it('should match snapshot', () => {
            expect(root).toMatchSnapshot();
        });    
    });
    describe('min', () => {
        beforeAll(()=>{
            m.mount(root, { view: () => m('.hs_white', [
                m(Pivot, <PivotAttrs>{ 
                      pivotName: 'by Area',
                      pivotHeaders: ['Sums', 'red', 'green', 'blue'],
                      table: {data: table, header:headers},
                      columns:[{'Value':'<color'}], // use `ColumnAccess` function and string column
                      by: ['area', 'shape']
                }),
             ])});
        });
    
        it('should match snapshot', () => {
            expect(root).toMatchSnapshot();
        });    
    });
    describe('count', () => {
        beforeAll(()=>{
            m.mount(root, { view: () => m('.hs_white', [
                m(Pivot, { 
                      pivotName: 'by Area',
                      table: {data: table, header:headers},
                      columns:[{'Value':'@color'}], // use `ColumnAccess` function and string column
                      by: ['area', 'shape']
                }),
             ])});
        });
    
        it('should match snapshot', () => {
            expect(root).toMatchSnapshot();
        });    
    });
});

