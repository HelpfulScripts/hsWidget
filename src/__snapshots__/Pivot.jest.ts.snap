// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pivot count should match snapshot 1`] = `
<div>
  <div
    class="hs_white"
  >
    <div
      class="hs_pivot"
    >
      <div
        class="row mon_header"
      >
        <span
          class="name"
        >
          by Area
        </span>
        <span
          class="hs_right"
        >
          blue
        </span>
        <span
          class="hs_right"
        >
          red
        </span>
        <span
          class="hs_right"
        >
          green
        </span>
      </div>
      <div
        class="hs_collapsible hs_collapsible_expanded pivot_row mon0"
      >
        <div
          class="hs_collapsible_title"
        >
          <div
            class="row"
          >
            <span
              class="name"
            >
               
            </span>
            <span
              class="hs_right"
            >
              3
            </span>
            <span
              class="hs_right"
            >
              1
            </span>
            <span
              class="hs_right"
            >
              2
            </span>
          </div>
        </div>
        <div
          class="hs_collapsible_content hs_scrolly"
          style="max-height: 400px; transition: max-height 0.25s ease-in-out;"
        >
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    N
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    1
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    NE
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    1
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    NW
                  </span>
                  <span
                    class="hs_right"
                  >
                    1
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    S
                  </span>
                  <span
                    class="hs_right"
                  >
                    2
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    SE
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    1
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pivot max and generator should match snapshot 1`] = `
<div>
  <div
    class="hs_white"
  >
    <div
      class="hs_pivot"
    >
      <div
        class="row mon_header"
      >
        <span
          class="name"
        >
          by Area
        </span>
        <span
          class="hs_right"
        >
          Sums
        </span>
        <span
          class="hs_right"
        >
          blue
        </span>
        <span
          class="hs_right"
        >
          red
        </span>
        <span
          class="hs_right"
        >
          green
        </span>
      </div>
      <div
        class="hs_collapsible hs_collapsible_expanded pivot_row mon0"
      >
        <div
          class="hs_collapsible_title"
        >
          <div
            class="row"
          >
            <span
              class="name"
            >
               
            </span>
            <span
              class="hs_right"
            >
              115
            </span>
            <span
              class="hs_right"
            >
              22
            </span>
            <span
              class="hs_right"
            >
              37
            </span>
            <span
              class="hs_right"
            >
              18
            </span>
          </div>
        </div>
        <div
          class="hs_collapsible_content hs_scrolly"
          style="max-height: 400px; transition: max-height 0.25s ease-in-out;"
        >
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    N
                  </span>
                  <span
                    class="hs_right"
                  >
                    37
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    37
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    NE
                  </span>
                  <span
                    class="hs_right"
                  >
                    18
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    18
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    NW
                  </span>
                  <span
                    class="hs_right"
                  >
                    17
                  </span>
                  <span
                    class="hs_right"
                  >
                    17
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    S
                  </span>
                  <span
                    class="hs_right"
                  >
                    34
                  </span>
                  <span
                    class="hs_right"
                  >
                    22
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    SE
                  </span>
                  <span
                    class="hs_right"
                  >
                    9
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    9
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pivot min should match snapshot 1`] = `
<div>
  <div
    class="hs_white"
  >
    <div
      class="hs_pivot"
    >
      <div
        class="row mon_header"
      >
        <span
          class="name"
        >
          by Area
        </span>
        <span
          class="hs_right"
        >
          blue
        </span>
        <span
          class="hs_right"
        >
          red
        </span>
        <span
          class="hs_right"
        >
          green
        </span>
      </div>
      <div
        class="hs_collapsible hs_collapsible_expanded pivot_row mon0"
      >
        <div
          class="hs_collapsible_title"
        >
          <div
            class="row"
          >
            <span
              class="name"
            >
               
            </span>
            <span
              class="hs_right"
            >
              12
            </span>
            <span
              class="hs_right"
            >
              37
            </span>
            <span
              class="hs_right"
            >
              9
            </span>
          </div>
        </div>
        <div
          class="hs_collapsible_content hs_scrolly"
          style="max-height: 400px; transition: max-height 0.25s ease-in-out;"
        >
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    N
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    37
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    NE
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    18
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    NW
                  </span>
                  <span
                    class="hs_right"
                  >
                    17
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    S
                  </span>
                  <span
                    class="hs_right"
                  >
                    12
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    SE
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    9
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pivot sum should match snapshot 1`] = `
<div>
  <div
    class="hs_white"
  >
    <div
      class="hs_pivot"
    >
      <div
        class="row mon_header"
      >
        <span
          class="name"
        >
          click to expand rows
        </span>
        <span
          class="hs_right"
        >
          sums
        </span>
        <span
          class="hs_right"
        >
          square
        </span>
        <span
          class="hs_right"
        >
          circle
        </span>
        <span
          class="hs_right"
        >
          triangle
        </span>
        <span
          class="hs_right"
        >
          cone
        </span>
      </div>
      <div
        class="hs_collapsible hs_collapsible_expanded pivot_row mon0"
      >
        <div
          class="hs_collapsible_title"
        >
          <div
            class="row"
          >
            <span
              class="name"
            >
               
            </span>
            <span
              class="hs_right"
            >
              37
            </span>
            <span
              class="hs_right"
            >
              35
            </span>
            <span
              class="hs_right"
            >
              49
            </span>
            <span
              class="hs_right"
            >
              22
            </span>
            <span
              class="hs_right"
            >
              9
            </span>
          </div>
        </div>
        <div
          class="hs_collapsible_content hs_scrolly"
          style="max-height: 400px; transition: max-height 0.25s ease-in-out;"
        >
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    blue
                  </span>
                  <span
                    class="hs_right"
                  >
                    22
                  </span>
                  <span
                    class="hs_right"
                  >
                    17
                  </span>
                  <span
                    class="hs_right"
                  >
                    12
                  </span>
                  <span
                    class="hs_right"
                  >
                    22
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    green
                  </span>
                  <span
                    class="hs_right"
                  >
                    18
                  </span>
                  <span
                    class="hs_right"
                  >
                    18
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    9
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
          <div>
            <div
              class="hs_collapsible  pivot_row mon1"
            >
              <div
                class="hs_collapsible_title"
              >
                <div
                  class="row"
                >
                  <span
                    class="name"
                  >
                    red
                  </span>
                  <span
                    class="hs_right"
                  >
                    37
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    37
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                  <span
                    class="hs_right"
                  >
                    
                  </span>
                </div>
              </div>
              <div
                class="hs_collapsible_content hs_scrolly"
                style="max-height: 0; transition: max-height 0.25s ease-in-out;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
